consumer/kafka/src/main/java/com/spotify/heroic/consumer/kafka/ConsumerThread.java

```diff
@@ -83,23 +82,38 @@ public ConsumerThread(
     public void run() {
         log.info("{}: Starting thread", name);
 
-        active.incrementAndGet();
+        threadIsStarting();
 
         try {
             guardedRun();
         } catch (final Throwable e) {
             log.error("{}: Error in thread", name, e);
-            active.decrementAndGet();
+            threadIsStopping();
             stopFuture.fail(e);
             return;
         }
 
         log.info("{}: Stopping thread", name);
-        active.decrementAndGet();
+        threadIsStopping();
         stopFuture.resolve(null);
         return;
     }
 
+    private void threadIsStarting() {
+        // Increment counter and report atomically
+        synchronized (active) {
```

Comment:
Very true, thanks for the pointer. Of course the synchronisation should be around the data and not at the caller. 
As you noted, I was coming from a mindset of keeping "active" in sync with the reporting.